pennylane/tape/tapes/jacobian_tape.py

```diff
@@ -438,34 +477,62 @@ def jacobian(self, device, params=None, **options):
                 # this computation is only performed once, for all parameters.
                 options["y0"] = np.asarray(self.execute_device(params, device))
 
-        jac = None
-        p_ind = range(len(params))
+        options["device"] = device
+
+        # collect all circuits (tapes) and postprocessing functions required
+        # to compute the jacobian
 
-        # loop through each parameter and compute the gradient
-        for idx, (l, param_method) in enumerate(zip(p_ind, allowed_param_methods)):
+        all_tapes = []
+        reshape_info = []
+        processing_fns = []
 
-            if param_method == "0":
-                # Independent parameter. Skip, as this parameter has a gradient of 0.
-                continue
+        for trainable_idx, param_method in enumerate(diff_methods):
 
             if (method == "best" and param_method[0] == "F") or (method == "numeric"):
                 # finite difference method
-                g = self.numeric_pd(l, device, params=params, **options)
+                tapes, processing_fn = self.numeric_pd(trainable_idx, params=params, **options)
 
             elif (method == "best" and param_method[0] == "A") or (method == "analytic"):
+                # some analytic methods need the device wires
+                options["dev_wires"] = device.wires
+
                 # analytic method
-                g = self.analytic_pd(l, device, params=params, **options)
+                tapes, processing_fn = self.analytic_pd(trainable_idx, params=params, **options)
+
+            # we create a flat list here to feed at once to the device
+            all_tapes.extend(tapes)
+            # to extract the correct result for this parameter later, remember the number of tapes
+            reshape_info.append(len(tapes))
+            processing_fns.append(processing_fn)
+
+        # execute all tapes
+        results = [
+            tape.execute(device) for tape in all_tapes
+        ]  # soon to be: device.batch_execute(all_tapes)
```

Comment:
Will this depend on whether the device inherits from `QubitDevice` or `Device` (looking at the differences in https://github.com/PennyLaneAI/pennylane/pull/844)?

Reply:
The current plan is to add the `batch_execute` function to both classes, so we can keep device logic out of here...

Reply:
Nice, thanks! I just saw the signatures were different so was wondering if that would get reflected here.